### PR TITLE
Implements Polkapool Pallet

### PR DESCRIPTION
**Pull request by Gauthamastro**
1. Implements the transaction store dispatchable function
2. Adds the required types to Config trait
3. Sets up the on_initialize function

Complete the following items to complete this PR

- [x] Remove `MinStakeAmount` from config trait 
The min stake amount should be determined using the weight of the dispatchable function, and then instead of `MinStakeAmount`, the config should define `MinStakePerWeight`. The min stake required is calculated by the weight of the transaction multiplied by `MinStakePerWeight`

- [x] Remove `MaxAllowedTxns` from config trait
The number of allowed transactions should be limited based on the transaction's weight instead of the number of txns. Limiting the number of txns based on the total weight gives a pre-determined resource usage in every block. Therefore, the config trait should define `MaxAllowedWeight`, which is defined as an `x%` (say 20% for testing) of the total block weight in runtime.

- [x] Similarly `MinStakePeriod` should also be removed for `MinStakePeriodPerWeight` ( up for discussion )
Here too, the total time a user should be staking the tokens depends on the total weight of the transaction multiplied by  `MinStakePeriodPerWeight`

Points to consider
 1. The goal is to make it costlier for an attacker to continue exploiting the system for feeless transactions.
 2. If it is based on total weight, heavier transactions will have a long stake period compared to lighter transactions, but this doesn't solve an attacker sending lots of lighter transactions.
 3.  We can implement a quadratic formula where the Stake period increases with more and more usage in a short period of time.

- [x] Implement a simple shuffler algorithm for shuffling a vector of calls based on the randomness from on-chain.

 Moved this task `Extend the `CallFilter` only to allow a pre-determined contract address from the contract's pallet ( up for discussion, is there a better way to filter feeless transactions) ` to #212 
- [x]  Governance dispatchable to slash the stake of defaulting users.

### CODE REVIEW

**Comment by ksr30** [comment]
@Gauthamastro are we going to merge this PR directly to the main?

**Comment by Gauthamastro** [comment]
> @Gauthamastro could you please verify the base branch ... I think it should be develop.

Thanks, @ksr30 for the heads up

**Comment by Gauthamastro** [comment]
we also a dispatchable function called unstake(origin) which will unlock the funds to the user if unlocking_block =< current_blocknumber

**Comment by mambisi** [comment]
I think using `MinStakePeriodPerWeight` might make user wait for too long before they can unstake their tokens because if the weight average call is 10_000 and `MinStakePeriodPerWeight` is 1, the user will have to wait 10_000 blocks before they can unstake

**Comment by Gauthamastro** [comment]
Ideally, we want users who are using too many feeless transactions to stake their tokens longer. At the same time, the heavier transaction needs to have a higher stake. 

StakePeriod should be a function of 
1. time period consecutive usage of feeless transactions
2. weight of the transaction

Is there anything more? @mambisi

**Comment by Gauthamastro** [comment]
We can use the fraction of the total allocated block weight for feeless for estimating the stake amount based the weight of the transaction

**Comment by mambisi** [comment]
` Extend the CallFilter only to allow a pre-determined contract address...` Outstanding waiting for implementation from @frankli-dev

**Comment by Gauthamastro** [comment]
> ` Extend the CallFilter only to allow a pre-determined contract address...` Outstanding waiting for implementation from @frankli-dev

Move the remaining task to a separate issue and close this one

**Comment by mambisi** [comment]
@Gauthamastro  Ready for review

File: pallets/polkapool/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        /// Returns staked tokens back to origin if `origin` is a staked user
+        #[weight = 10000]
+        pub fn unstake(origin) -> DispatchResult {
+            let who = ensure_signed(origin.clone())?;
+            ensure!(origin.clone().into().is_ok(),Error::<T>::BadOrigin);
+            ensure!(<StakedUsers<T>>::contains_key(&who),Error::<T>::StakeNotFound);
+            let stake = <StakedUsers<T>>::get(&who);
+            let current_block_no: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
+            ensure!(stake.unlocking_block <= current_block_no,Error::<T>::UnlockingFailedCurrentBlockNumberLow);
+            ensure!(T::Currency::deposit(AssetId::POLKADEX, &who, stake.staked_amount).is_ok(),Error::<T>::FailedToDepositStakedAmount);
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:313]:
```suggestion
            ensure!(T::Currency::unreserve(AssetId::POLKADEX, &who, stake.staked_amount).is_ok(),Error::<T>::FailedToDepositStakedAmount);
```

File: pallets/polkapool/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+            ensure!( call_weight < u32::MAX as u64, Error::<T>::Overflow);
+            // Store the staking record
+            let mut staked_amount = Self::staked_users(who.clone());
+            staked_amount.staked_amount += stake_amount;
+
+            //Todo change to new Formula
+            staked_amount.unlocking_block =  staked_amount.unlocking_block
+            .checked_add(&(call_weight as u32).saturated_into::<T::BlockNumber>().checked_mul(&T::MinStakePeriodPerWeight::get().saturated_into::<T::BlockNumber>()).ok_or(<Error<T>>::Overflow)?)
+            .ok_or(<Error<T>>::Overflow)?;
+
```
Comment (Gauthamastro) [inline pallets/polkapool/src/lib.rs:287]: Add `T::Currency::reserve()` and reserve the staked_amount,

**Review by mambisi — APPROVED** [review]

**Pull request merged**